src/com/ms/silverking/cloud/dht/net/IPAliasingUtil.java

```diff
@@ -0,0 +1,149 @@
+      IpAliasConfiguration ipAliasConfiguration = mc.getIpAliasConfiguration(dhtConfig.getIpAliasMapName());
+      return readAliases(dhtConfig, ipAliasConfiguration);
+    } catch (KeeperException | IOException ex) {
+      Log.severe("Failed to lookup aliases on ZK, will try locally", ex);
+      return new IPAliasMap(readFileAliases(config.getPort()));
+    }
+  }
+
+  static void logAliasMap(Map<IPAndPort, IPAndPort[]> aliasMap) {
+    logAliasMap(aliasMap, Level.INFO);
+  }
+
+  static void logAliasMap(Map<IPAndPort, IPAndPort[]> aliasMap, Level level) {
+    if (aliasMap != null && !aliasMap.isEmpty()) {
+      StringBuilder logMsg = new StringBuilder();
+      logMsg.append("Aliases for nodes will be used: [");
+      for (Map.Entry<IPAndPort, IPAndPort[]> entry : aliasMap.entrySet()) {
+        StringBuilder addressList = new StringBuilder();
+        for (IPAndPort addr : entry.getValue()) {
+          addressList.append(addr);
+        }
+        logMsg.append(String.format("%s --> %s; ", entry.getKey(), addressList.toString()));
+      }
+      logMsg.append("]");
+      Log.log(level, logMsg.toString());
+    }
+  }
+
+  static Map<IPAndPort, IPAndPort[]> parseConfigAliasMap(Map<String, String> aliasMap, int configPort) {
+    if (aliasMap != null) {
+      ImmutableMap.Builder<IPAndPort, IPAndPort[]> builder = ImmutableMap.builder();
+
+      for (Map.Entry<String, String> entry : aliasMap.entrySet()) {
+        String ip = entry.getKey();
+        if (ip.contains(":")) {
+          throw new RuntimeException("lhs must be daemon ip only, not daemon ip:port");
+        }
+        String aliasToken = entry.getValue();
+        builder.put(new IPAndPort(ip, configPort), parseAliasToken(aliasToken, configPort));
+      }
+      return builder.build();
+    } else {
+      return readFileAliases(configPort);
+    }
+  }
+
+  /**
+   * Reads the IP alias map from a file declared by DHTConstants.ipAliasMapFileVar envVar
+   * File structure - ip+ip,ip,ip - one set per line. The LHS is the real IP, the RHS is a set of aliases
+   * OR ip+ip:port,ip:port,ip:port..
+   * N.b. if port is not provided in the alias, a fixed port is assumed as declared in the DHT config
+   *
+   * @param fixedPort the fixed port expected for the SK cluster (from DHT config)
+   * @return a map of ip -> List of aliases
+   */
+  private static Map<IPAndPort, IPAndPort[]> readFileAliases(int fixedPort) {
+    String aliasMapFile;
+
+    aliasMapFile = DHTConstants.defaultDefaultClassVars.getVarMap().get(DHTConstants.ipAliasMapFileVar);
+    Log.warningf("%s=%s", DHTConstants.ipAliasMapFileVar, aliasMapFile);
+    if (aliasMapFile != null && aliasMapFile.trim().length() > 0) {
+      try {
+        FileInputStream in = new FileInputStream(new File(aliasMapFile));
+        BufferedReader reader;
+        String line;
+        Map<IPAndPort, IPAndPort[]> map;
+
+        map = new HashMap<>();
+        reader = new BufferedReader(new InputStreamReader(in));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced in statements at the following lines: 107, 109. The resource closure statement is at line: 116. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Either a) close *reader* in a try-finally block or b) close the resource by declaring *reader* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).